src/java/org/apache/cassandra/service/ConsensusTableMigrationState.java

```diff
@@ -0,0 +1,697 @@
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.Config.LegacyPaxosStrategy;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.dht.IPartitionerDependentSerializer;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.repair.RepairJobDesc;
+import org.apache.cassandra.repair.RepairResult;
+import org.apache.cassandra.repair.messages.RepairOption;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.paxos.Paxos;
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.tcm.ClusterMetadataService;
+import org.apache.cassandra.tcm.Epoch;
+import org.apache.cassandra.tcm.transformations.BeginConsensusMigrationForTableAndRange;
+import org.apache.cassandra.tcm.transformations.MaybeFinishConsensusMigrationForTableAndRange;
+import org.apache.cassandra.tcm.transformations.SetConsensusMigrationTargetProtocol;
+import org.apache.cassandra.utils.NullableSerializer;
+import org.apache.cassandra.utils.PojoToString;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.apache.cassandra.db.TypeSizes.sizeof;
+import static org.apache.cassandra.dht.Range.intersectionOfNormalizedRanges;
+import static org.apache.cassandra.dht.Range.normalize;
+import static org.apache.cassandra.dht.Range.rangeSerializer;
+import static org.apache.cassandra.dht.Range.subtract;
+import static org.apache.cassandra.dht.Range.subtractNormalizedRanges;
+import static org.apache.cassandra.utils.CollectionSerializers.deserializeMap;
+import static org.apache.cassandra.utils.CollectionSerializers.deserializeSet;
+import static org.apache.cassandra.utils.CollectionSerializers.newCollectionSerializer;
+import static org.apache.cassandra.utils.CollectionSerializers.newHashMap;
+import static org.apache.cassandra.utils.CollectionSerializers.serializeCollection;
+import static org.apache.cassandra.utils.CollectionSerializers.serializeMap;
+import static org.apache.cassandra.utils.CollectionSerializers.serializedCollectionSize;
+import static org.apache.cassandra.utils.CollectionSerializers.serializedMapSize;
+
+/**
+ * Track and update the migration state of individual table and ranges within those tables
+ */
+public abstract class ConsensusTableMigrationState
```

Comment:
Let's make an `o.a.c.service.consensus` package, and possibly an `o.a.c.service.consensus.migration` package, and move some of this stuff in there

Reply:
What would go into the `consensus` package besides the `migration` sub package? Multi-word package names aren't common, but nesting an empty directory also seems weird.